rclcpp/src/rclcpp/subscription_base.cpp

```diff
@@ -301,6 +301,25 @@ SubscriptionBase::can_loan_messages() const
   return rcl_subscription_can_loan_messages(subscription_handle_.get());
 }
 
+bool
+SubscriptionBase::take_loaned_message(
```

Comment:
I believe that it must return the loaned message to the middleware, unless that will be memory leak or probably middleware cannot reuse the memory for the publisher side. probably message type should be templated and `LoanedMessage` object should be returned, so that it can return the loaned memory to the middleware via dtor?

Reply:
sorry for the delay, so to clarify you think we should just using a `LoanedMessage` object instead of the `void * loaned_message`? This then allows us to keep the rmw middleware in the function - `take_loaned_message` but allow the `LoanedMessage loaned_message` to deallocate the message in it's dtor [here](https://github.com/ros2/rclcpp/blob/rolling/rclcpp/include/rclcpp/loaned_message.hpp#L156).

Originally when I made this pull request I was trying to match the [design docs for loaned messages](http://design.ros2.org/articles/zero_copy.html) where the user could either:
> - Manage the memory for the message instance themselves, providing a reference to it, into which the middleware should fill the data.
> - Take one or more loaned messages from the middleware and return the loans later.

I assumed the rclcpp API would just act as a wrapper to rcl, and would just have this object as a placeholder, which intends zero-copy.

Reply:
Additionally while I think `rclcpp::LoanedMessage` is much better than a `void * loaned_message`, `LoanedMessage` is currently based on the `Publisher` side of things:
```
  LoanedMessage(
    const rclcpp::PublisherBase & pub,
    std::allocator<MessageT> allocator)
```
Also see [TODO](https://github.com/ros2/rclcpp/blob/rolling/rclcpp/src/rclcpp/executor.cpp#L631-L632), therefore we would have to revamp the LoanedMessage class to support subscriptions. I wouldn't mind working on supporting subscriptions in that class, especially since I see multiple issues and PRs based on `LoanedMessage` and intra-process messaging. Should we implement `Subscription` in `LoanedMessage` first, or should we push what we have then move to that?

Reply:
@fujitatomoya The change here does not affect the return of loanedbuf to the middleware.It is just act as a wrapper to rcl,code logic is the same as before

Reply:
@sgf201, I don't think this PR is where it needs to be to be honest. I need to take a bit more time to refine these changes to a better level. I'm going to make this a draft until I have it at the point where is needs to be.